File: platform/fastapi_server/config/lifecycle/20_cache_service.lifecycle.py
"""
Cache Service Lifecycle Hook for FastAPI

Initializes the multi-instance cache factory and stores it in app.state
for use in routes and dependencies.

Loading Order: 20 (after config, before business logic)

Usage in routes:
    from fastapi import Request, Depends
    from typing import Annotated

    # Direct access
    @app.get('/token')
    async def get_token(request: Request):
        providers = request.app.state.cache.get(CacheNames.PROVIDERS)
        token = await providers.get_or_set('oauth:google', fetch_token, 600)
        return {'token': token}

    # Or with dependency injection
    def get_providers_cache(request: Request):
        return request.app.state.cache.get(CacheNames.PROVIDERS)

    @app.get('/token')
    async def get_token(cache: Annotated[ICacheService, Depends(get_providers_cache)]):
        token = await cache.get_or_set('oauth:google', fetch_token, 600)
        return {'token': token}
"""

import logging
import sys
from pathlib import Path
from typing import Any
from fastapi import FastAPI

logger = logging.getLogger("lifecycle:cache_service")

# Add polyglot package to path
polyglot_path = Path(__file__).parent.parent.parent.parent / "polyglot" / "server_provider_cache" / "py"
if str(polyglot_path) not in sys.path:
    sys.path.insert(0, str(polyglot_path))

from server_provider_cache import (
    create_cache_factory,
    CacheNames,
    DefaultTTLs,
)
from env_resolver import resolve_redis_env


def get_cache_config(app: FastAPI) -> dict[str, Any]:
    """
    Get cache configuration from app config or environment variables.

    Args:
        app: FastAPI application instance

    Returns:
        Cache configuration dictionary
    """
    # Try to get from app config first
    cache_config: dict[str, Any] = {}

    if hasattr(app.state, 'config') and hasattr(app.state.config, 'get'):
        try:
            cache_config = app.state.config.get('cache') or {}
            logger.debug("Loaded cache config from app.state.config")
        except Exception:
            logger.warning("No cache config in app config, using defaults")

    default_ttl = cache_config.get('defaultTtl') or cache_config.get('default_ttl')
    if default_ttl is None:
        _redis_env = resolve_redis_env()
        default_ttl = _redis_env.cache_default_ttl
        logger.debug("Using CACHE_DEFAULT_TTL from env: %d", default_ttl)

    return {
        'default_ttl': default_ttl,
        'backend': cache_config.get('backend') or resolve_redis_env().cache_backend,
        # Per-cache TTL overrides
        'ttls': {
            'providers': cache_config.get('ttls', {}).get('providers', DefaultTTLs.PROVIDERS),
            'services': cache_config.get('ttls', {}).get('services', DefaultTTLs.SERVICES),
            'config': cache_config.get('ttls', {}).get('config', DefaultTTLs.CONFIG),
            'sessions': cache_config.get('ttls', {}).get('sessions', DefaultTTLs.SESSIONS),
            'tokens': cache_config.get('ttls', {}).get('tokens', DefaultTTLs.TOKENS),
        },
    }


async def onStartup(app: FastAPI, config: dict[str, Any]) -> None:
    """
    Startup hook - Initialize cache factory and create default cache instances.

    Args:
        app: FastAPI application instance
        config: Bootstrap config (may be overridden by app config)
    """
    logger.info("Starting cache_service lifecycle hook...")
    try:
        logger.info("Initializing Cache Service...")

        cache_config = get_cache_config(app)

        logger.info(
            "Cache configuration: default_ttl=%d, backend=%s",
            cache_config["default_ttl"], cache_config["backend"],
        )
        logger.debug("Cache TTLs: %s", cache_config["ttls"])

        # Create the factory with defaults
        factory = create_cache_factory(
            defaults={
                'backend': cache_config['backend'],
                'default_ttl': cache_config['default_ttl'],
            }
        )
        logger.debug("Cache factory created")

        # Create default cache instances based on CacheNames
        # These can be customized via app config
        factory.create(CacheNames.PROVIDERS, default_ttl=cache_config['ttls']['providers'])
        factory.create(CacheNames.SERVICES, default_ttl=cache_config['ttls']['services'])
        factory.create(CacheNames.CONFIG, default_ttl=cache_config['ttls']['config'])
        factory.create(CacheNames.SESSIONS, default_ttl=cache_config['ttls']['sessions'])
        factory.create(CacheNames.TOKENS, default_ttl=cache_config['ttls']['tokens'])

        # Store in app state
        app.state.cache = factory
        app.state.CacheNames = CacheNames

        logger.info(
            "Cache service initialized: %s (%d instances)",
            factory.get_names(), factory.get_count(),
        )
        logger.info("cache_service lifecycle hook completed successfully")
    except Exception as exc:
        logger.error("cache_service lifecycle hook failed: %s", exc, exc_info=True)
        raise


async def onShutdown(app: FastAPI, config: dict[str, Any]) -> None:
    """
    Shutdown hook - Cleanup cache resources.

    Args:
        app: FastAPI application instance
        config: Server configuration dictionary
    """
    logger.info("Starting cache_service shutdown...")
    try:
        if hasattr(app.state, 'cache') and hasattr(app.state.cache, 'destroy_all'):
            logger.info("Shutting down cache service...")
            await app.state.cache.destroy_all()
            logger.info("Cache service shut down successfully")
        else:
            logger.debug("No cache service to shut down")
    except Exception as exc:
        logger.error("cache_service shutdown failed: %s", exc, exc_info=True)
        raise

File: platform/platform-core/fastify/src/lifecycles/20_cache_service.lifecycle.mjs
/**
 * Cache Service Lifecycle Hook for Fastify
 *
 * Initializes the multi-instance cache factory and decorates the server
 * with the cache factory instance for use in routes and plugins.
 *
 * Loading Order: 20 (after config, before business logic)
 *
 * Usage in routes:
 *     // Get factory
 *     const factory = req.server.cache;
 *
 *     // Get specific cache instance
 *     const providers = req.server.cache.get(CacheNames.PROVIDERS);
 *     const token = await providers.getOrSet('oauth:google', fetchToken, 600);
 *
 *     // Or use convenience method
 *     const token = await req.server.cache.get(CacheNames.PROVIDERS).getOrSet(...);
 */

import {
    createCacheFactory,
    CacheNames,
    DefaultTTLs,
} from '../../../polyglot/server_provider_cache/mjs/index.mjs';
import { resolveRedisEnv } from '@internal/env-resolver';

/**
 * Get cache configuration from app config or environment variables.
 *
 * @param {object} server - Fastify server instance
 * @returns {object} Cache configuration
 */
function getCacheConfig(server) {
    // Try to get from app config first
    let cacheConfig = {};

    if (server.config && typeof server.config.get === 'function') {
        try {
            cacheConfig = server.config.get('cache') || {};
        } catch (e) {
            server.log.debug('No cache config in app config, using defaults');
        }
    }

    const redisEnv = resolveRedisEnv();
    return {
        defaultTtl: cacheConfig.defaultTtl
            || redisEnv.cacheDefaultTtl,
        backend: cacheConfig.backend
            || redisEnv.cacheBackend,
        // Per-cache TTL overrides
        ttls: {
            providers: cacheConfig.ttls?.providers || DefaultTTLs[CacheNames.PROVIDERS],
            services: cacheConfig.ttls?.services || DefaultTTLs[CacheNames.SERVICES],
            config: cacheConfig.ttls?.config || DefaultTTLs[CacheNames.CONFIG],
            sessions: cacheConfig.ttls?.sessions || DefaultTTLs[CacheNames.SESSIONS],
            tokens: cacheConfig.ttls?.tokens || DefaultTTLs[CacheNames.TOKENS],
        },
    };
}

/**
 * Startup hook - Initialize cache factory and create default cache instances.
 *
 * @param {object} server - Fastify server instance
 * @param {object} config - Bootstrap config (may be overridden by app config)
 */
export async function onStartup(server, config) {
    server.log.info('[lifecycle:cache-service] Initializing Cache Service...');

    try {
        const cacheConfig = getCacheConfig(server);

        server.log.info({
            defaultTtl: cacheConfig.defaultTtl,
            backend: cacheConfig.backend,
        }, '[lifecycle:cache-service] Cache configuration resolved');

        server.log.debug({ ttls: cacheConfig.ttls }, '[lifecycle:cache-service] Per-cache TTL overrides');

        // Create the factory with defaults
        server.log.info('[lifecycle:cache-service] Creating cache factory');
        const factory = createCacheFactory({
            defaults: {
                backend: cacheConfig.backend,
                defaultTtl: cacheConfig.defaultTtl,
            },
        });

        // Create default cache instances based on CacheNames
        // These can be customized via app config
        server.log.info('[lifecycle:cache-service] Creating default cache instances');
        factory.create(CacheNames.PROVIDERS, {
            defaultTtl: cacheConfig.ttls.providers,
        });
        factory.create(CacheNames.SERVICES, {
            defaultTtl: cacheConfig.ttls.services,
        });
        factory.create(CacheNames.CONFIG, {
            defaultTtl: cacheConfig.ttls.config,
        });
        factory.create(CacheNames.SESSIONS, {
            defaultTtl: cacheConfig.ttls.sessions,
        });
        factory.create(CacheNames.TOKENS, {
            defaultTtl: cacheConfig.ttls.tokens,
        });

        // Decorate server with cache factory
        if (!server.hasDecorator('cache')) {
            server.decorate('cache', factory);
            server.log.info('[lifecycle:cache-service] Decorated server with cache factory');
        } else {
            server.cache = factory;
            server.log.info('[lifecycle:cache-service] Overwrote existing server.cache with new factory');
        }

        // Also expose CacheNames for convenience
        if (!server.hasDecorator('CacheNames')) {
            server.decorate('CacheNames', CacheNames);
            server.log.info('[lifecycle:cache-service] Decorated server with CacheNames');
        }

        // Register cleanup hook
        server.addHook('onClose', async () => {
            server.log.info('[lifecycle:cache-service] Cleaning up cache service...');
            await factory.destroyAll();
            server.log.info('[lifecycle:cache-service] Cache service cleaned up');
        });

        server.log.info({
            caches: factory.getNames(),
            count: factory.getCount(),
        }, '[lifecycle:cache-service] Cache service initialized successfully');
    } catch (err) {
        server.log.error({ err, hookName: '20-cache-service' }, '[lifecycle:cache-service] Cache service initialization failed');
        throw err;
    }
}

/**
 * Shutdown hook - Cleanup cache resources.
 *
 * Note: This is also handled by the onClose hook registered in onStartup,
 * but this provides an explicit shutdown point if needed.
 *
 * @param {object} server - Fastify server instance
 */
export async function onShutdown(server) {
    if (server.cache && typeof server.cache.destroyAll === 'function') {
        server.log.info('Shutting down cache service...');
        await server.cache.destroyAll();
    }
}
